compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java

```diff
@@ -631,6 +627,21 @@ private boolean recordFieldsAssignableToMap(BRecordType recordType, BMapType tar
         return recordType.fields.stream().allMatch(field -> isAssignable(field.type, targetMapType.constraint));
     }
 
+    private boolean isAssignableMapType(BMapType sourceMapType, BRecordType targetRecType) {
+        if (targetRecType.sealed) {
+            return false;
+        }
+
+        for (BField field : targetRecType.fields) {
+            if (!Symbols.isFlagOn(field.symbol.flags, Flags.OPTIONAL) ||
```

Comment:
!(isOptional && isAsignable)
Shall we refactor this use use `and` instead of `or`